### PR TITLE
bt-19: Sign-in route

### LINKED ISSUE #19: Sign In
- [x] Repository (Model) (getting user for sign-in)
- [x] Service (sign-in user flow)
- [x] Controller (sign-in route)
- [x] Validation (shared)

### PR DESCRIPTION
**Pull request by ptarho**
New PR without merged bt-86 branch. 
I commented places that have to be replaced with bcrypt. 
Currently it`s just set password as "HASH" and compare user input with that value

### CODE REVIEW

File: backend/src/bundles/users/user.entity.ts [hunk 1]
```diff
@@ -65,10 +65,12 @@ class UserEntity implements Entity {
     public toObject(): {
         id: number;
         email: string;
+        passwordHash: string;
```
Comment (nikita-remeslov) [inline backend/src/bundles/users/user.entity.ts]: token should not be part of user entity

File: shared/src/bundles/users/types/user-find-response-dto.type.ts [hunk 2]
```diff
@@ -1,6 +1,7 @@
 type UserFindResponseDto = {
     id: number;
     email: string;
+    passwordHash: string;
```
Comment (nikita-remeslov) [inline shared/src/bundles/users/types/user-find-response-dto.type.ts]: hash should not be a part of response

File: backend/src/bundles/auth/auth.service.ts [hunk 3]
```diff
@@ -12,17 +16,60 @@ class AuthService {
...
+                cause: 'Email already used',
+                status: HttpCode.CONFLICT,
+            });
+        }
+
         const user = await this.userService.create(userRequestDto);
-        return this.signIn(user.toObject().id);
+        return {
+            token: await tokenService.create({ id: user.id }),
+        };
```
Comment (nikita-remeslov) [inline backend/src/bundles/auth/auth.service.ts:45]:
```suggestion
        return {
            token,
            user,
        };
```

File: backend/src/bundles/auth/auth.service.ts [hunk 4]
```diff
@@ -12,17 +16,60 @@ class AuthService {
...
     }
 
-    public async signIn(id: number): Promise<{ token: string }> {
+    public async signIn(
+        userRequestDto: UserSignInRequestDto,
+    ): Promise<UserSignInResponseDto> {
+        const user = await this.verifyLoginCredentials(userRequestDto);
         return {
-            token: await tokenService.create({ id }),
+            token: await tokenService.create({ id: user.id }),
```
Comment (nikita-remeslov) [inline backend/src/bundles/auth/auth.service.ts]:
```suggestion
            token,
            user,
```

Comment (nikita-remeslov) [reply backend/src/bundles/auth/auth.service.ts]: you should return user and token

File: backend/src/bundles/auth/auth.service.ts [hunk 5]
```diff
@@ -12,17 +16,60 @@ class AuthService {
...
 
     public async signUp(
         userRequestDto: UserSignUpRequestDto,
     ): Promise<UserSignUpResponseDto> {
+        const { email } = userRequestDto;
+
+        const userByEmail = await this.userService.findByEmail(email);
+        if (userByEmail) {
+            throw new HttpError({
+                message: 'User already exist.',
```
Comment (nikita-remeslov) [inline backend/src/bundles/auth/auth.service.ts]: add to error-message.enum.ts

File: backend/src/bundles/auth/auth.service.ts [hunk 6]
```diff
@@ -12,17 +16,60 @@ class AuthService {
...
+                cause: 'Email already used',
```
Comment (nikita-remeslov) [inline backend/src/bundles/auth/auth.service.ts]:
```suggestion
```

File: backend/src/bundles/auth/auth.service.ts [hunk 7]
```diff
@@ -12,17 +16,60 @@ class AuthService {
...
+                status: HttpCode.CONFLICT,
```
Comment (nikita-remeslov) [inline backend/src/bundles/auth/auth.service.ts]: should be 400 not 409

File: backend/src/bundles/auth/auth.service.ts [hunk 8]
```diff
@@ -12,17 +16,60 @@ class AuthService {
...
+    public async verifyLoginCredentials({
+        email,
+        password,
+    }: UserSignInRequestDto): Promise<UserFindResponseDto> {
+        const user = await this.userService.findByEmail(email);
+
+        if (!user) {
+            throw new HttpError({
+                message: 'User not found',
+                cause: 'User with the given email does not exist',
```
Comment (nikita-remeslov) [inline backend/src/bundles/auth/auth.service.ts]:
```suggestion
```

File: backend/src/bundles/auth/auth.service.ts [hunk 9]
```diff
@@ -12,17 +16,60 @@ class AuthService {
...
+
+    public async verifyLoginCredentials({
+        email,
+        password,
+    }: UserSignInRequestDto): Promise<UserFindResponseDto> {
+        const user = await this.userService.findByEmail(email);
+
+        if (!user) {
+            throw new HttpError({
+                message: 'User not found',
```
Comment (nikita-remeslov) [inline backend/src/bundles/auth/auth.service.ts]: add to shared

File: backend/src/bundles/auth/auth.service.ts [hunk 10]
```diff
@@ -12,17 +16,60 @@ class AuthService {
...
+                cause: 'User with the given email does not exist',
+                status: HttpCode.NOT_FOUND,
+            });
+        }
+
+        const isEqualPassword = password === 'HASH'; // Replace with cryptCompare from bt-86
+
+        if (!isEqualPassword) {
+            throw new HttpError({
+                message: 'Unauthorized',
```
Comment (nikita-remeslov) [inline backend/src/bundles/auth/auth.service.ts]: add to shared

File: backend/src/bundles/auth/auth.service.ts [hunk 11]
```diff
@@ -12,17 +16,60 @@ class AuthService {
...
+                cause: 'Invalid credentials',
```
Comment (nikita-remeslov) [inline backend/src/bundles/auth/auth.service.ts]:
```suggestion
```

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

File: backend/src/bundles/users/user.service.ts [hunk 12]
```diff
@@ -43,14 +43,18 @@ class UserService implements Service {
...
+    public async create(
+        payload: UserSignUpRequestDto,
+    ): Promise<UserFindResponseDto> {
+        const passwordHash = 'HASH'; // Remove in bt-86 with encrypt
+        const user = await this.userRepository.create(
             UserEntity.initializeNew({
                 email: payload.email,
-                passwordSalt: 'SALT', // TODO
-                passwordHash: 'HASH', // TODO
+                passwordSalt: 'SALT', // Remove with a new migration ticket
```
Comment (nikita-remeslov) [inline backend/src/bundles/users/user.service.ts]: already in the development

File: backend/src/bundles/auth/auth.service.ts [hunk 13]
```diff
@@ -12,17 +16,60 @@ class AuthService {
...
+                message: 'User already exist.',
+                cause: 'Email already used',
+                status: HttpCode.CONFLICT,
+            });
+        }
+
         const user = await this.userService.create(userRequestDto);
-        return this.signIn(user.toObject().id);
+        return {
+            token: await tokenService.create({ id: user.id }),
```
Comment (nikita-remeslov) [inline backend/src/bundles/auth/auth.service.ts]:
```suggestion
            token,
	        user,
```

**Comment by nikita-remeslov** [comment]
conflicts

**Comment by ptarho** [comment]
@nikita-remeslov fixed

File: backend/src/bundles/auth/auth.service.ts (hunk 7 above)
Comment (nikita-remeslov) [reply backend/src/bundles/auth/auth.service.ts]:
```suggestion
                status: HttpCode.BAD_REQUEST,
```

**Review by nikita-remeslov — CHANGES_REQUESTED** [review]

**Review by nikita-remeslov — APPROVED** [review]

**Review by MaksymGavatuk — APPROVED** [review]

**Review by Monyster — APPROVED** [review]

**Pull request merged**